README.md
Pratt Parser
============

[![Gem Version](https://badge.fury.io/rb/pratt_parser.svg)](http://badge.fury.io/rb/pratt_parser)
[![Build Status](https://travis-ci.org/tommay/pratt_parser.svg)](https://travis-ci.org/tommay/pratt_parser)

A Pratt Parser.  Just a simple parsing engine.

Define tokens that describe your language and combine terms.  Write a
Lexer (an Enumerator) that produces a stream of tokens.  Instantiate a
PrattParser and call #eval to parse the stream and return whatever
your tokens want.

Pratt parsers are like recursive descent parsers but instead of having
to code a function for each production you just code up some token
objects which define their own prededence and associativity and how
subexpressions to the left and/or right are combined.  So it's simple
to add new language features/tokens without having to rewrite a bunch
of recursive descent functions.

Pratt parsers are also more efficient than recursive descent parsers
since they don't need to recurse all the way down to the bottom level
to figure out what to do.

Read more:

http://javascript.crockford.com/tdop/tdop.html

http://effbot.org/zone/simple-top-down-parsing.htm

http://journal.stuffwithstuff.com/2011/03/19/pratt-parsers-expression-parsing-made-easy/


CHANGELOG
## 0.1.0 ##

  * Add postfix factorial operator to examples.
  * PrattParser#expect and #if? can now take a block so token
    expectations can be based on something more general than class.

## 0.0.3 ##

  * examples/tree.rb supports prefix + and -.
  * gem.required_ruby_version is >= 1.9.

## 0.0.2 ##

  * Add examples/tree.rb.

## 0.0.1 ##

  * Include examples.

## 0.0.0 ##

  * Initial release.
		


spec/pratt_parser_spec.rb
# Need a lot more tests than this but it's a start.

require "rubygems"
require "bundler/setup"
require "minitest/autorun"
require "pratt_parser"

describe PrattParser do
#  it "parses an empty expression" do
#    lexer = Enumerator.new {}
#    PrattParser.new(lexer).eval.must_be nil
#  end

  it "calls nud(self) for the first token in a (sub)expression" do
    token = MiniTest::Mock.new
    lexer = Enumerator.new do |y|
      y << token
    end
    parser = PrattParser.new(lexer)
    token.expect(:nud, "ok", [parser])
    parser.eval.must_equal "ok"
    token.verify
  end

  it "passes the lookahead token to an expect block" do
    expected_token = Object.new

    token_a = Object.new.tap do |o|
      o.singleton_class.class_eval do
        define_method(:nud) do |parser|
          parser.expect do |token|
            token.must_be_same_as expected_token
          end
        end
      end
    end

    lexer = Enumerator.new do |y|
      y << token_a
      y << expected_token
    end

    PrattParser.new(lexer).eval
  end

  it "moves on to the next token on a valid expect" do
    class TokenA
      def nud(parser)
        parser.expect(TokenB)
        "A"
      end
    end
    class TokenB; end
    class TokenC
      def lbp; 0; end
    end

    expected_token = MiniTest::Mock.new
    expected_token.expect(:class, TokenB, [])

    fetched_expected_token = false
    lexer = Enumerator.new do |y|
      y << TokenA.new
      y << expected_token
      fetched_expected_token = true
      y << TokenC.new
    end

    PrattParser.new(lexer).eval

    assert fetched_expected_token
  end

  it "moves on to the next token on a valid expect with a block" do
    class TokenA
      def nud(parser)
        parser.expect {}
      end
    end
    class TokenC
      def lbp; 0; end
    end

    fetched_expected_token = false
    lexer = Enumerator.new do |y|
      y << TokenA.new
      y << Object.new
      fetched_expected_token = true
      y << TokenC.new
    end

    PrattParser.new(lexer).eval

    assert fetched_expected_token
  end

  it "fails on invalid expect" do
    class TokenA
      def nud(parser)
        parser.expect(TokenC)
        "A"
      end
    end
    class TokenB; end
    class TokenC; end

    unexpected_token = MiniTest::Mock.new
    unexpected_token.expect(:class, TokenB, [])
    unexpected_token.expect(:class, TokenB, [])

    lexer = Enumerator.new do |y|
      y << TokenA.new
      y << unexpected_token
    end

    assert_raises RuntimeError do
      PrattParser.new(lexer).eval
    end
  end

  it "fails on invalid expect with a block" do
    class TokenA
      def nud(parser)
        parser.expect do
          raise "That's not what I was expecting"
        end
      end
    end

    lexer = Enumerator.new do |y|
      y << TokenA.new
      y << Object.new
    end

    assert_raises RuntimeError do
      PrattParser.new(lexer).eval
    end
  end

end


lib/pratt_parser.rb
# A Pratt parser.  Similar to a recursive descent parser but instead of
# coding a function for each production, the syntax is coded in a set
# of token objects that are yielded by the lexer.  New operators and
# statements can be slipped into the language with the proper
# precedence by adding new token objects to the lexer without altering
# the code for existing tokens.  Pretty cool.
#
# lexer is must have an +#each+ method that returns token objects.  A token
# has three methods:
# +lbp+::
#   Returns the operator precedence.  Higher numbers bind more tightly.
# <tt>nud(parser)</tt>::
#   Called when the token is the first token in an expression,
#   including a recursive call to +expresssion+ (i.e., subexpression).
#   For example, +nud+ would be called for a unary operator, a literal,
#   or for the "if" in the construct "if <cond> then <expr>".  It is
#   the token's responsibility to call +parser.expression+,
#   +parser.expect+, and/or +parser.if?+ to handle the remainder of the
#   (sub)expression, if any.
# <tt>led(parser, left)</tt>::
#   Called when the token is preceeded by a subexpression, passed in
#   as +left+.  The token may be postfix or infix.  It is the token's
#   responsibility to call +parser.expression+, +parser.expect+,
#   and/or +parser.if?+ to handle the remainder of the expression, if
#   any, and combine it with +left+.
#
# Only +lbp+ is mandatory.  +nud+ and +led+ will be called only when
# necessary, if ever.  For example, +nud+ will never be called for a
# strictly infix token. If the token appears at the start of a
# (sub)expression then an exception that isn't at al appropriate
# to the abstraction will be thrown.
#
# +nud+ and +led+ can call <tt>parser.expression(rbp)</tt> to recursively
# parse the right subexpression.  +rbp+ should be the token's +lbp+ for
# left-associativity, +lbp-1+ for right.
#
# <tt>PrattParser.new(lexer).eval</tt> will return the result of the parse.
#
# Syntax errors aren't handled at the moment and will cause ridiculous
# exceptions to be raised such as +NoMethodError+.
#
# Further reading:
# * http://javascript.crockford.com/tdop/tdop.html
# * http://effbot.org/zone/simple-top-down-parsing.htm
# * http://journal.stuffwithstuff.com/2011/03/19/pratt-parsers-expression-parsing-made-easy/

class PrattParser
  # Creates a new +PrattParser+.  +lexer+ is an +Enumerable+, or something
  # with an +#each+ method.
  #
  def initialize(lexer)
    @lexer = Enumerator.new do |y|
      lexer.each do |token|
        y << token
      end
      y << EndToken.new
    end

    @token = nil
  end

  # Runs the tokens through the parse engine and returns the result,
  # or throws some exception on parse error.
  #
  def eval
    @token = @lexer.next
    expression(0)
  end

  # For use by token +#led+ methods to parse subexpressions with
  # binding power less than +rbp+.  Whatever that means.  Don't worry,
  # it just does the Right Thing.
  #
  def expression(rbp)
    t = @token
    @token = @lexer.next
    left = t.nud(self)
    while rbp < @token.lbp
      t = @token
      @token = @lexer.next
      left = t.led(self, left)
    end
    left
  end

  # Checks whether the lookahead token is of the
  # +expected_token_class+ and raises an exception if it isn't.
  # Alternatively a block may be given; the block is passed the
  # lookahead token and should raise an exception if it's not an
  # expected token.  In either case if no exception is raised then the
  # lookahead token is consumed.
  #
  # +expect+ can be used to match the
  # right parenthesis in a parenthesized expression, the colon in a
  # "cond ? then : else" expression, etc.
  #
  def expect(expected_token_class = nil, &block)
    block ||= lambda do |token|
      if token.class != expected_token_class
        raise "Expected #{expected_token_class}, got #{token.class}"
      end
    end
    block.call(@token)
    @token = @lexer.next
  end

  # Checks whether the lookahead token is of the
  # +token_class+.  If it is, consumes the lookahead token
  # and returns truthy, else just returns falsy.
  # Alternatively a block can be given which is passed the token
  # and should return truthy or falssy.
  #
  # +if?+ can be used for optional tokens such as the +else+
  # clause in "if cond then val1 [else val2] end".
  #
  def if?(token_class = nil, &block)
    block ||= lambda do |token|
      token.class == token_class
    end
    if block.call(@token)
      @token = @lexer.next
    end
  end

  # A private class.  An instance of +EndToken+ is automatically
  # appended as the last token in the input stream.  With its +lbp+ of
  # zero it will terminate the parse.
  #
  class EndToken
    def lbp                     # :nodoc:
      0
    end
  end
end


examples/expression_evaluator.rb
#!/usr/bin/env ruby

# Evaluate simple arithmetic expressions using a PrattParser.
# Supports +, -, *, /, ^ (with customary precedence and
# associativity), and parentheses.  Also supports = to compare numbers
# and booleans.  Adding support for < and > would be a simple matter
# of adding some more tokens.
#
# Supporting whitespace needs a lexer that throws away whitespace.
#
# Numeric integer constants are supported using single-digit tokens
# which are left-associative.

require "pratt_parser"

class PrattEvaluator
  def self.eval(expression)
    PrattParser.new(Lexer.new(expression)).eval
  end

  class Lexer
    # Note that new returns an Enumerator, not a Lexer.

    def self.new(expression)
      expression.each_char.lazy.map{|c|@@tokens[c]}
    end

    class Token
      def initialize(lbp)
        @lbp = lbp
      end

      def lbp
        @lbp
      end
    end

    class InfixToken < Token
      def initialize(lbp, associates = :left, &block)
        super(lbp)
        @block = block
        @rbp = (associates == :left ? lbp : lbp - 1)
      end

      def led(parser, left)
        @block.call(left, parser.expression(@rbp))
      end
    end

    class PostfixToken < Token
      def initialize(lbp, &block)
        super(lbp)
        @block = block
      end

      def led(parser, left)
        @block.call(left)
      end
    end

    class DigitToken < Token
      def initialize(lbp, value)
        super(lbp)
        @value = value
      end
      
      def nud(parser)
        @value
      end
      
      def led(parser, left)
        left*10 + @value
      end
    end
    
    class LeftParenToken < Token
      def nud(parser)
        parser.expression(lbp).tap do
          parser.expect(RightParenToken)
        end
      end
    end
    
    class RightParenToken < Token
    end
    
    @@tokens = {}
    
    def self.token(char, t)
      @@tokens[char] = t
    end

    def self.infix(char, lbp, associates = :left, &block)
      token(char, InfixToken.new(lbp, associates, &block))
    end

    def self.postfix(char, lbp, &block)
      token(char, PostfixToken.new(lbp, &block))
    end

    def self.factorial(n)
      n <= 1 ? 1 : n * factorial(n - 1)
    end

    token("(", LeftParenToken.new(1))
    token(")", RightParenToken.new(1))

    infix("=", 10, &:==)
    infix("+", 20, &:+)
    infix("-", 20, &:-)
    infix("*", 30, &:*)
    infix("/", 30, &:/)
    infix("^", 40, :right, &:**)
    postfix("!", 50) {|left| factorial(left)}

    (0..9).each do |d|
      token(d.to_s, DigitToken.new(100, d.to_f))
    end
  end
end

if __FILE__ == $0
  puts PrattEvaluator.eval(ARGV[0])
end


examples/tree.rb
#!/usr/bin/env ruby

# Parses simple arithmetic expressions using a PrattParser.  Supports
# +, -, *, /, ^, ?: (with customary precedence and associativity), and
# parentheses.  + and - may be either prefix or infix.  The parse
# returns a tree of *Node objects which is printed prefix-style ala
# Lisp.
#
# Numeric integer constants are supported using single-digit tokens
# which are left-associative.

require "bundler/setup"
require "pratt_parser"

class UnaryNode
  def initialize(operator, node)
    @operator = operator
    @node = node
  end

  def to_s
    "(#{@operator} #{@node})"
  end

  # Returns a pretty-print version of the expression ala Lisp.

  def pp(indent = "")
    newindent = indent + "  "
    "#{indent}(#{@operator}\n#{@node.pp(newindent)})"
  end
end

class BinaryNode
  def initialize(operator, left, right)
    @operator = operator
    @left = left
    @right = right
  end

  def to_s
    "(#{@operator} #{@left} #{@right})"
  end

  # Returns a pretty-print version of the expression ala Lisp.

  def pp(indent = "")
    newindent = indent + "  "
    "#{indent}(#{@operator}\n#{@left.pp(newindent)}\n#{@right.pp(newindent)})"
  end
end

class TernaryNode
  def initialize(operator, cond, if_expr, else_expr)
    @operator = operator
    @cond = cond
    @if_expr = if_expr
    @else_expr = else_expr
  end

  def to_s
    if @else_expr
      "(#{@operator} #{@cond} #{@if_expr} #{@else_expr})"
    else
      "(#{@operator} #{@cond} #{@if_expr})"
    end
  end

  # Returns a pretty-print version of the expression ala Lisp.

  def pp(indent = "")
    newindent = indent + "  "
    if @else_expr
      "#{indent}(#{@operator}\n#{@cond.pp(newindent)}\n#{@if_expr.pp(newindent)}\n#{@else_expr.pp(newindent)})"
    else
      "#{indent}(#{@operator}\n#{@cond.pp(newindent)}\n#{@if_expr.pp(newindent)})"
    end
  end
end

class NumberNode
  def initialize(number)
    @number = number
  end

  def number
    @number
  end

  def to_s
    @number.to_s
  end

  # Returns a pretty-print version of the expression ala Lisp.

  def pp(indent = "")
    "#{indent}#{@number}"
  end
end

class TreeBuilder
  def self.eval(expression)
    PrattParser.new(Lexer.new(expression)).eval
  end

  class Lexer
    # Note that new returns an Enumerator, not a Lexer.

    def self.new(expression)
      Enumerator.new do |y|
        expression.scan(%r{\b(?:if|then|else|end)\b|\S}) do |token|
          y << @@tokens[token]
        end
      end
    end

    class Token
      def initialize(lbp)
        @lbp = lbp
      end

      def lbp
        @lbp
      end
    end

    class InfixToken < Token
      def initialize(operator, lbp, associates = :left)
        super(lbp)
        @operator = operator
        @rbp = (associates == :left ? lbp : lbp - 1)
      end

      def led(parser, left)
        right = parser.expression(@rbp)
        BinaryNode.new(@operator, left, right)
      end
    end

    # A Bifix token can be either prefix (via nud) or infix (via lcd).

    class BifixToken < InfixToken
      def nud(parser)
        # Bind super-tight to the right.
        right = parser.expression(1000000)
        UnaryNode.new(@operator, right)
      end
    end

    class PostfixToken < Token
      def initialize(function, lbp)
        super(lbp)
        @function = function
      end

      def led(parser, left)
        UnaryNode.new(@function, left)
      end
    end

    class DigitToken < Token
      def initialize(lbp, value)
        super(lbp)
        @value = value
      end
      
      def nud(parser)
        NumberNode.new(@value)
      end

      def led(parser, left)
        NumberNode.new(left.number*10 + @value)
      end
    end
    
    class LeftParenToken < Token
      def nud(parser)
        parser.expression(lbp).tap do
          parser.expect(RightParenToken)
        end
      end
    end
    
    class RightParenToken < Token
    end
    
    class QuestionToken < Token
      def led(parser, cond)
        if_expr = parser.expression(lbp)
        parser.expect(ColonToken)
        else_expr = parser.expression(lbp)
        TernaryNode.new("?", cond, if_expr, else_expr)
      end
    end

    class ColonToken < Token
    end

    class IfToken < Token
      def nud(parser)
        cond = parser.expression(lbp)
        parser.expect{|token| token.expect("then")}
        then_expr = parser.expression(lbp)
        if parser.if?{|token| token.if?("else")}
          else_expr = parser.expression(lbp)
        end
        parser.expect{|token| token.expect("end")}
        TernaryNode.new("if", cond, then_expr, else_expr)
      end
    end

    class KeywordToken < Token
      def initialize(lbp, keyword)
        super(lbp)
        @keyword = keyword
      end

      def expect(keyword)
        if keyword != @keyword
          raise "Expected #{keyword} token, got #{@keyword}"
        end
      end

      def if?(keyword)
        keyword == @keyword
      end
    end
    
    @@tokens = {}
    
    def self.token(char, t)
      @@tokens[char] = t
    end

    def self.infix(char, lbp, associates = :left)
      token(char, InfixToken.new(char, lbp, associates))
    end

    def self.bifix(char, lbp)
      token(char, BifixToken.new(char, lbp, :left))
    end

    def self.postfix(char, lbp, function, &block)
      token(char, PostfixToken.new(function, lbp))
    end

    token("(", LeftParenToken.new(1))
    token(")", RightParenToken.new(1))

    infix("=", 10)
    token("if", IfToken.new(12))
    token("then", KeywordToken.new(12, "then"))
    token("else", KeywordToken.new(12, "else"))
    token("end", KeywordToken.new(12, "end"))
    token("?", QuestionToken.new(15))
    token(":", ColonToken.new(15))
    bifix("+", 20)
    bifix("-", 20)
    infix("*", 30)
    infix("/", 30)
    infix("^", 40, :right)
    postfix("!", 50, "factorial")

    (0..9).each do |d|
      token(d.to_s, DigitToken.new(100, d))
    end
  end
end

if __FILE__ == $0
  tree = TreeBuilder.eval(ARGV[0])
  puts tree.to_s
  puts tree.pp
end
